pkg/kubelet/kubelet.go

```diff
@@ -2952,6 +2969,87 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
 	}, 100*time.Millisecond, stopCh)
 }
 
+// handleMaxCheckpointsPerContainer goes through the list of checkpoints of a
+// container in a pod and deletes old checkpoints if there are more than
+// 'kl.maxCheckpointsPerContainer'.
+func (kl *Kubelet) handleMaxCheckpointsPerContainer(podUID types.UID, podFullName, containerName string, options *runtimeapi.CheckpointContainerRequest) error {
+
+	if kl.maxCheckpointsPerContainer <= 0 {
+		// This cannot really happen as the kubelet config parser will
+		// error out early if this is <= 0. This additional error
+		// handling is mainly here to make sure tests are working.
+		return errors.New("maxCheckpointsPerContainer must be greater than 0")
+	}
+
+	// Lock while accessing the checkpoints on disk to avoid chaos
+	// if multiple threads would do it at the same time.
+	kl.maxCheckpointsPerContainerLock.Lock()
+	defer kl.maxCheckpointsPerContainerLock.Unlock()
+
+	checkpointDirectoryPath := filepath.Join(
+		kl.getCheckpointsDir(),
+		string(podUID),
```

Comment:
I believe podUID an be reused here..

Reply:
If the namespace, pod and container have the same name or one of the described collisions above. Thanks to the timestamp and the counter it should not result in checkpoints being overwritten, but maybe older checkpoints, from a podUID collisions, being removed first. Not sure if that is a problem or not.

Reply:
yeah was more a tree/identifier issue... here with poduid.. one will need to tree | grep to find their checkpoints by name, here you can also identify by poduid.. if that was nec.. and missing we could add the poduid to the original name as an alternative

Reply:
The name of the checkpoint is returned when triggering the kubelet checkpoint API endpoint. I am also working on code to see how it could look like if the kubelet checkpoint API endpoint is available on the kubectl level.

What I am trying to say is that searching for a checkpoint archive is something I would not expect that happens a lot, so that the path is not super important. Whatever makes most sense works for me. The main goal is to avoid any ambiguity in the file name. The podUID was suggested somewhere along the review in this PR and makes sense to me.

There was also agreement that we can change the location as long this is still marked as an alpha feature.

As mentioned in another comment using https://github.com/checkpoint-restore/checkpointctl is something I would recommend to get all details from the checkpoint archive and rely less on encoding information into the file name.

Reply:
> The pain point this PR is addressing is over use of the drive.. which should not be happening in the approved forensics cases?

It feels like this seems to be one of our main discussion points. I agree that it "should not be happening" but at the same time it is something people are actively talking about. From my point of view it feels unrealistic to expect that people always clean up unused files. That is why I want to help by cleaning it up automatically.

> At min, IMO, we should tell sig-node about this change to begin automatically removing checkpoints past some limit and get a general consensus if that is acceptable in the interim before the next checkpoint KEP update.

Okay. I will add it on the sig-node agenda.

Reply:
I think if we want to introduce this clean up functionality we need to add it to the existing KEP as an additional feature or start a new one. As discussed at SIG Node meeting today, once we have this setting we can start exploring integrating this into GC and eviction logic, having per-pod policies, etc. This all creates an unnecessary burden on a kubelet.